### Title: 2023_10_23_LSG_polyseq Analysis
### Author: Kate Johnson
### Date: 11/01/23
### Add'tl Notes:

# setup
libraries
```{r}
# all-purpose data analysis tools
library(tidyverse)
# ggplot extension
library(cowplot)
#correlation plots
library(corrplot)
# easier string formatting
library(glue)
# filepath manipulation
library(fs)

# edgeR and limma for DE analysis
library(edgeR)
library(limma)

# for reading/writing .gct files
#library(cmapR)

# for working with matrices (especially sparse ones)
library(Matrix)

#complex heatmap
library(ComplexHeatmap)
library(circlize) #for colorRamp2 color scaling
```
directories
```{r}
wd = getwd()
# For accessing shared data on server
shared_dir = "/Volumes/nchevrier/katej/projects/polysome-seq/2023_10_23_LSG_polyseq/bcbio"
# For reading in counts/metadata
input_dir = path(wd, "../input/data/2_formatted_data")
# For saving plots  
output_dir = path(wd, "../output/2023_10_30-LSG")
```

sources
```{r}

```

#complex heatmap plot fxn
move to plotting functions!
```{r}
do_plot = function(file_name, w, h, p) {
  pdf(file=path(output_dir, file_name), width = w, height = h)
      draw(p)
      dev.off()
}
```

# Read in Data
```{r, warnings =F}

#counts
counts = read_csv(path(input_dir, "20231101_LSG_polysome_counts.csv")) 
#rename gene column
colnames(counts)[1] = "gene"
# format matrix to be ready for de analysis
genes <- counts$gene
counts$gene <- NULL
rownames(counts) <- genes

#meta        
metadata = read_csv(path(input_dir, "20231101_LSG_polysome_metadata.csv")) %>%
           #drop unused first col
           select(-1)
```
# QC

complexity
- ratio of unique molecules to the number of reads
- unique molecular identifiers (UMIs) allow us to collapse molecules that we read multiple times
    - A higher complexity means more information for the amount of reads we have
- calculate complexity by comparing the UMI counts (`counts`) to the uncollapsed read counts 
- should be ~0.6, so a large deviation below that might be cause for concern
- raw read counts are called "tagcounts-dupes" by bcbio

```{r}
## make a table of total counts per sample
counts_summary = enframe(colSums(counts), name='Name', value='counts') %>%
    left_join(metadata)

# read in raw read counts
counts_dir = path(shared_dir, "final/2023-10-31_bcbio")

# read matrix
reads = readMM(path(counts_dir, 'tagcounts-dupes.mtx'))

# add rownames and colnames
rownames(reads) = read_lines(path(counts_dir, 'tagcounts-dupes.mtx.rownames'))
colnames(reads) = read_lines(path(counts_dir, 'tagcounts-dupes.mtx.colnames'))

## calculate counts for reads and add to summary table
# get read counts
reads = enframe(colSums(reads), name = 'sample', value = 'read_counts') %>%
        rename_all(~str_to_title(.))

# join to counts
counts_summary = left_join(counts_summary, reads)

#plot
p<-counts_summary %>%
    mutate(complexity = counts / Read_counts) %>%
    ggplot(aes(str_c(pool, "_", replicate),
               complexity, fill=pool)) + 
        geom_col() + 
        facet_grid(. ~ treatment) + 
        labs(x = "Polysome Fraction", title = "Sample Complexity (UMIs / Uncollapsed Read Counts)") +
        #guides(fill = 'none') + 
        theme(axis.text.x = element_text(angle = 45, hjust = .5, vjust = .5, size = rel(0.5)),
              plot.title = element_text(hjust = 0.5))

#save_plot
save_plot(path(output_dir, "QC_plots/sample_complexity_replicate_barplot.pdf"),p)

```

counts per sample
```{r}
p<-ggplot(counts_summary, aes(str_c(pool, '_', replicate), 
                   counts, fill=pool)) + 
    geom_col() + 
    facet_grid(treatment ~ .) + 
    #scale_y_log10() +
    #guides(fill = 'none') + 
    xlab('Polysome Fraction') +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5))

save_plot(path(output_dir, "QC_plots/counts_per_sample.pdf"), p, dpi = 90)
```

replicate correlation
```{r}
# commenting out for now bc not particularly informative
#pairwise correlations between samples for each treatment
#for(treatment in unique(metadata$treatment)){
#    suppressWarnings(
#        gplots::heatmap.2(cor(counts[, metadata$treatment == treatment]),
#                          trace = 'none', Rowv=F, Colv=F, 
#                         col=viridis::viridis(75),
#                          margins = c(10, 10))
#    )
#}

# table of all pairwise correlations < 0.9
# get all pairwise correlations
cormelt = as_tibble(reshape2::melt(cor(counts)))
colnames(cormelt) = c('s1', 's2', 'cor')

filt_cormelt = cormelt %>%
    # extract pieces of name for sample 1
    extract(s1, c('treatment', 'replicate1', 'pool'), '(.*)_(.*)_(.*)') %>%
    # extract pieces of name for sample 2
    extract(s2, c('treatment2', 'replicate2', 'pool2'), '(.*)_(.*)_(.*)') %>%
    # only keep samples where treatment and pool match
    # replicate1 < replicate2 removes duplicate comparisons
    filter(treatment == treatment2, pool == pool2, replicate1 < replicate2) %>%
    # comparison column makes for easier reading
    mutate(comparison = str_c(replicate1, '-', replicate2)) %>%
    # remove extraneous columns
    select(-treatment2, -pool2, -replicate1, -replicate2) #%>%
    # all correlations are above 0.9 so keep everything for now
    # filter(cor < .9)


p<-ggplot(filt_cormelt,
       aes(x = treatment, y = pool, fill = cor)) +
       geom_tile() +
       geom_text(aes(label = round(cor,3)), 
                  hjust = 0.5, vjust = 0.5) +
       facet_wrap(~comparison)
  
#save plot 
save_plot(path(output_dir, "QC_plots/replicate_correlation.pdf"),p, dpi = 90)
```


# Data Normalization
```{r}
#set cpm thresholds (keep low to start)
cpm_threshold = 10
cpm_needed = 2

groups <- factor(str_c(metadata$treatment, '_', metadata$pool))
dge = DGEList(counts = counts, group = groups)
dge_norm = calcNormFactors(dge)
isexpr = rowSums(cpm(dge_norm) > cpm_threshold) >= cpm_needed

dge_norm_fltd = dge_norm[isexpr, ]

cpm = cpm(dge_norm_fltd) #dropped to ~8,000 genes
cpm_all = cpm(dge_norm)

# reorder cols so its easier to open in morpheus
# use metadata to help establish order
metadata = metadata %>%
           arrange(treatment, pool, replicate)
cpm = cpm[,metadata$Name]

#save cpm heatmaps for plotting in morpheus
write.csv(as.matrix(cpm), path(output_dir, "cpm_tables/20231101_LSG_cpmTable_filt.csv"))
write.csv(as.matrix(cpm_all), path(output_dir, "cpm_tables/20231101_LSG_cpmTable_all.csv"))

```

PCA
```{r}
# run pca
pca = prcomp(t(cpm), center=T, scale=T)
# combine pca results with metadata for easy plotting
pca_res = bind_cols(metadata, as_tibble(pca$x))

# plot

# color by treatment
p1<-ggplot(pca_res, aes(PC1, PC2, color=treatment)) + 
    geom_point(size=2) +
    scale_color_manual(values = c("grey", "peachpuff1","#F68064"))

# color by pool
p2<-ggplot(pca_res, aes(PC1, PC2, color=pool)) + 
    geom_point(size=2) +
    scale_color_manual(values = c("#68C3A5", "#B9CEE6", "#8C9FCA","#707FA1"))

# color by replicate
p3<-ggplot(pca_res, aes(PC1, PC2, color=replicate)) + 
    geom_point(size=2)

# save plots
save_plot(path(output_dir, "QC_plots/PCA/PCA_by_treatment.pdf"),p1,dpi=90)
save_plot(path(output_dir, "QC_plots/PCA/PCA_by_pool.pdf"),p2,dpi=90)
save_plot(path(output_dir, "QC_plots/PCA/PCA_by_replicate.pdf"),p3,dpi=90)
```


# DE analysis

The general steps for modeling are: 
1. Filter cpm and metadata based on the tissue at hand
2. Make a design matrix specifying the experimental condition of each sample 
3. Run `voom()` to transform and prepare data
4. Run initial `lmFit()` to fit data
5. Setup contrasts between groups for tests you want to run
    1. List contrasts
    2. Make contrast matrix
6. Run `contrast.fit()` to fit contrasts
7. Run `eBayes` to estimate statistic results
8. Gather results into useful tables

```{r}
# FILTER COUNTS & METADATA TO DROP BAD REPLICATE----
# LSGh6 input rep1 has low counts and clusters poorly -- so we will drop it
#set cpm thresholds (keep low to start)
cpm_threshold = 10
cpm_needed = 2

counts_filt = counts %>%
              select(-LSG6h_Rep1_Input)
meta_filt = metadata %>% filter(Name %in% colnames(counts_filt))
rownames(counts_filt) = rownames(counts)

# check that col order of counts_filt and row order of meta_filt match
sum(colnames(counts_filt) == meta_filt$Name) #sum is 35 (same as so looks like all match)

# re-do dge list object creation
groups <- factor(str_c(meta_filt$treatment, '_', meta_filt$pool))
dge = DGEList(counts = counts_filt, group = groups)
dge_norm = calcNormFactors(dge)
isexpr = rowSums(cpm(dge_norm) > cpm_threshold) >= cpm_needed
dge_norm_fltd = dge_norm[isexpr, ]


# DESIGN MATRIX----
# create factor with all conditions in experiment
f <- factor(groups, levels=unique(groups))

# build a design matrix for modeling with limma voom and indicating condition for each sample
design <- model.matrix(~0 + f)

# renames column names in the design (remove extra "f" at beginning of each group name)
colnames(design) <- sub("f", "", colnames(design))
rownames(design) <- colnames(counts_filt)

##LIMMA VOOM 

# limma voom transforms count data to log2-counts per million (log2-cpm) with associated weights
# estimate the mean-variance relationship and use this to compute appropriate observational-level weights.
# The data are then ready for linear modeling.
y <- voom(dge_norm_fltd, design, plot=T)

# fit linear model for each gene given a series of arrays
fit <- lmFit(y, design)

##CONTRAST MATRIX----
# each contrast for DE will be one pool in treated - control (ie. LSG2h_input - Control_input)
# 1. get all treatments that arent control
# 2. make an equivalent list with the same pools in control
# 3. subtract 1 - 2
all_cond = levels(groups)[!grepl("Control", (levels(groups)))]
conts = gsub(".*_", "Control_", all_cond)
contrasts_list = str_c(all_cond, " - ", conts)

contrast.matrix <- makeContrasts(contrasts = contrasts_list, levels = design)

#do some renaming of scores to facilitate pivoting later on and make our lives easier...
contrast.matrix = contrast.matrix %>%
                  as.data.frame %>% #convert to df for now for manipulation
                  rename_with(., ~gsub(" - Control.*", "", .)) %>%
                  as.matrix()

fit.cont <- contrasts.fit(fit, contrast.matrix)

e.fit <- eBayes(fit.cont)

#first bind together all LFC lists for all genes from ea. contrast
log2FC.all = as.data.frame(map_dfc(1:length(colnames(e.fit$contrasts)),
                             function(coef) topTable(e.fit, coef=coef, number=Inf, adjust = "BH",
                                                     sort.by='none')$logFC)) %>%
      data.frame()

adj_p = as.data.frame(map_dfc(1:length(colnames(e.fit$contrasts)),
                                    function(coef) topTable(e.fit, coef=coef, adjust = "BH",
                                                            number=Inf,
                                                            sort.by='none')$adj.P.Val))

rownames(log2FC.all) = rownames(adj_p) = make.unique(rownames(dge_norm_fltd), sep = ".")
colnames(log2FC.all) = colnames(adj_p) = colnames(e.fit$contrasts)

#turn log2FC.all.binary into matrix of -1,0,1
FDR_thresh = 0.01
log2FC.all.binary = adj_p %>%
              mutate_all(~ifelse(. < FDR_thresh, 1, 0))

#filter LFC matrix accordingly
log2FC.all.binary[(log2FC.all < 0) == T] = -1 * log2FC.all.binary[(log2FC.all < 0) == T]
names = rownames(log2FC.all.binary)
log2FC.all.binary = log2FC.all.binary %>%
               data.frame()

DE_log2FC.all = log2FC.all[rowSums(log2FC.all.binary != 0) > 0,] %>%
          data.frame()
rownames(DE_log2FC.all) = rownames(log2FC.all)[rowSums(log2FC.all.binary != 0) > 0]

#filter DE matrix based on LFC thresh
LFC_thresh = 1
DE_LFC_thresh = DE_log2FC.all %>%
          filter_all(any_vars(abs(.) > LFC_thresh))

#write csvs of DEGs and all log2FC
write.csv(as.matrix(DE_LFC_thresh), path(output_dir, "DE_analysis/filtered/FDR0.1_LFC1/20231101_filtered_FDR0.1_LFC1_DE-genes-lfcTable.csv"))

```

# Translation Efficiency ----
First plot % of total transcripts in each pool (take cpm in each pool and divide by input cpm)
Use only DE genes 
We will use this to help assess different methods of calculating TE
- First use FDR < 0.01 |LFC| > 0 (~7304 genes)
```{r}

#bind rownames as its own column ('gene') to make sure gene names dont get lost during manipulation
percent_transcript = cbind(gene = rownames(DE_log2FC.all), cpm[rownames(DE_log2FC.all),]) %>%
                     #make df for easier manipulation
                     data.frame() %>%
                     #pivot longer so we can average over replicates
                     pivot_longer(cols = -gene, names_to ="Name", values_to = "cpm") %>%
                     #split "Name" column into relevant info
                     separate(Name, into = c("Treatment", "Rep", "Pool"), sep="_") %>%
                     #pivot wider to get reps in separate cols
                     pivot_wider(names_from = "Rep", values_from = "cpm") %>%
                     #make cpm numeric (preserving NA values)
                     mutate_at(.vars = vars(Rep1:Rep3), ~ifelse(!(is.na(.)), as.numeric(.), .)) %>%
                     #calculate average
                        #struggling to get rowwise operations working correctly
                        #so sticking to a more manual calc for now
                     mutate(Avg_cpm = (Rep1 + Rep2 + Rep3) / 3) %>%
                     #drop raw info after spot checking averages
                     select(-contains("Rep"))
                    
#store inputs separately
inputs = percent_transcript %>%
         filter(Pool == "Input") %>%
         #drop Treatment column to avoid duplicates during merge
         select(-Pool) %>%
         rename("Avg_input_cpm" = Avg_cpm)

#bind inputs back to other pools and divide by inputs to get %
percent_total = percent_transcript %>%
                     #filter out inputs
                     filter(Pool != "Input") %>%
                     # bind inputs
                     merge(inputs, by = c("gene", "Treatment")) %>%
                     #calculate % of input 
                     # because we have normalized some ratios will be > 100%
                     mutate(percent_of_total = Avg_cpm/Avg_input_cpm * 100)

#format df for complex heatmap
df = percent_total %>%
     select(-contains("Avg")) %>%
     pivot_wider(names_from = c("Treatment", "Pool"),
                 values_from = "percent_of_total")

#prepare annotations
top_annot = data.frame(Name = colnames(df)[-1]) %>%
            separate(Name, into = c("Treatment", "Pool"), sep ="_")
rownames(top_annot) = colnames(df)[-1]

rwnmes = df$gene
H = df %>%
    select(-gene) %>%
    as.matrix()
rownames(H) = rwnmes
  
p<-Heatmap(H,
        column_title = "% of Total mRNA in Pool",
        name = "Avg_pool_cpm / Avg_input_cpm",
        show_row_names = F,
        show_column_names=F,
        #cluster_rows =T,
        cluster_columns = F,
        top_annotation = HeatmapAnnotation(df = top_annot,
                         col = list(Treatment = c(Control = "grey",
                                                   LSG2h = "#F89F7C",
                                                  LSG6h = "#F68064"),
                                   Pool = c("Pool1" = "#68C3A5", 
                                            "Pool2" ="#539C84",
                                            "Pool3" ="#3E7563")),
                          border = T,
                          show_legend = F,
                          annotation_name_gp = gpar(fontsize = 8)),
        column_split = top_annot$Treatment,
        column_gap = unit(2, "mm"),
        row_km = 20,
        row_gap = unit(0, "mm"),
        col = colorRamp2(c(0, 100, 200), c("white", "#8C9FCA", "darkblue"))
        )

#save plot
do_plot( "TE_analysis/Percent-mRNA-in-pool/FDR0.01/Pool-cpm_div_input-cpm.pdf", 8, 6, p)

#write csv
write.csv(as.matrix(H), path(output_dir, "TE_analysis/Percent-mRNA-in-pool/FDR0.01/Pool_cpm-div-Input_cpm.csv"))


#same plot but only 3,000 highly regulated transcripts

H_filt = H[rownames(DE_LFC_thresh),]
H_filt_scale = H_filt %>%
               t() %>%
               scale() %>%
               t()

p<-Heatmap(H_filt_scale,
        column_title = "% of Total mRNA in Pool",
        name = "Avg_pool_cpm / Avg_input_cpm",
        show_row_names = F,
        show_column_names=F,
        #cluster_rows =T,
        cluster_columns = F,
        top_annotation = HeatmapAnnotation(df = top_annot,
                         col = list(Treatment = c(Control = "grey",
                                                   LSG2h = "#F89F7C",
                                                  LSG6h = "#F68064"),
                                   Pool = c("Pool1" = "#68C3A5", 
                                            "Pool2" ="#539C84",
                                            "Pool3" ="#3E7563")),
                          border = T,
                          show_legend = F,
                          annotation_label = c("",""),
                          annotation_name_gp = gpar(fontsize = 8)),
        column_split = top_annot$Pool,
        column_gap = unit(2, "mm"),
        row_km = 30,
        row_gap = unit(0, "mm"),
        col = colorRamp2(c(min(H_filt_scale),-1, 0, max(H_filt_scale)), 
                         c("white", "white","#8C9FCA","darkblue"))
        )

#save plot
do_plot( "TE_analysis/Percent-mRNA-in-pool/FDR0.01/Pool-cpm_div_input-cpm_LFC_1-relative.pdf", 8, 6, p)

#write csv
write.csv(as.matrix(H_filt), path(output_dir, "TE_analysis/Percent-mRNA-in-pool/FDR0.01/Pool_cpm-div-Input_cpm_LFC_1.csv"))


####---- percent usage side by side with LFC
H2 = DE_LFC_thresh  
top_annot2 <- data.frame(Name = colnames(H2)) %>%
              separate(Name, into = c("Treatment", "Pool"), sep = "_")
rownames(top_annot2) = colnames(H2)


H2_scale = H2 %>%
           t() %>%
           scale() %>%
           t()

p2 <-Heatmap(H2_scale,
        column_title = "LFC",
        name = "Log2(Treated/Control)",
        show_row_names = F,
        show_column_names=F,
        #cluster_rows =T,
        cluster_columns = F,
        top_annotation = HeatmapAnnotation(df = top_annot2,
                         col = list(Treatment = c(Control = "grey",
                                                   LSG2h = "#F89F7C",
                                                  LSG6h = "#F68064"),
                                   Pool = c("Input" = "white",
                                            "Pool1" = "#68C3A5", 
                                            "Pool2" ="#539C84",
                                            "Pool3" ="#3E7563")),
                          border = T,
                          show_legend = T,
                          annotation_name_gp = gpar(fontsize = 8)),
        column_split = top_annot2$Treatment,
        column_gap = unit(2, "mm"),
        row_km = 30,
        row_gap = unit(0, "mm"),
        col = colorRamp2(c(-3,0,3), 
                         c("blue", "white","red"))
        )

p3 <- p+p2

#save plot for now (re-order clusters later)
do_plot( "TE_analysis/Percent-mRNA-in-pool/FDR0.01/LFC_1/PErcent-in-pool_PLUS_logfoldchange.pdf", 
         8, 6, p3)
```

#Re-design contrasts matrix
First approach: only perform comparisons in contrasts matrix
```{r}
##CONTRAST MATRIX----
# each contrast for DE will be one pool in treated - control (ie. LSG2h_input - Control_input)
# 1. get all treatments that arent control
# 2. make an equivalent list with the same pools in control
# 3. subtract 1 - 2
all_cond = levels(groups)[!grepl("Control", (levels(groups)))]
conts = gsub(".*_", "Control_", all_cond)
contrasts_list = str_c(all_cond, " - ", conts)

#DEFINE ADDT'L CONTRASTS FOR TE
# think about ways to streamline this
# set up df of contrasts to help with easy addition/subtraction
TE_contrasts = data.frame(contrasts = str_c("(", contrasts_list, ")")) %>%
               mutate(Treatment = gsub("\\(", "", gsub("_.*", "",contrasts)),
                      Pool = gsub(".*_", "", gsub(" - .*", "", contrasts))) %>%
               pivot_wider(names_from = "Treatment", values_from = "contrasts") %>%
               t() %>% #creates matrix
               data.frame() %>%
               #hack for combining strings from different columns
               mutate(pseudo_free = str_c("(", select(., X1) %>%
                                            reduce(str_c, sep =""),
                                             ") - (",
                                             select(., X2:X4) %>% 
                                            reduce(str_c, sep=" + "), 
                                             ")")) %>%
               #update naming for delta_Input column (change in mRNA = delta_mRNA)
               rename("delta.mRNA" = X1,
                      "L" = X2,
                      "M" = X3,
                      "H" = X4) %>%
               #create TE columns
               mutate(TE = str_c("(", 
                                 select(., M:H) %>% 
                                 reduce(str_c, sep=" + "), 
                                 ") - (", 
                                 select(., L) %>%
                                 reduce(paste0),
                                 ")"),
                      delta.M = str_c("(", 
                                 select(., M) %>% 
                                 reduce(paste0), 
                                 ") - (", 
                                 select(., L) %>%
                                 reduce(paste0),
                                 ")"),
                      delta.H = str_c("(", 
                                 select(., H) %>% 
                                 reduce(paste0), 
                                 ") - (", 
                                 select(., L) %>%
                                 reduce(paste0),
                                 ")"))

#write csv of contrasts scheme for future reference
write.csv(as.matrix(TE_contrasts), path(output_dir, "TE_analysis/TE_contrasts_matrix.csv"))

#pivot contrasts table to get ready for named vector to feed into makeContrasts
TE_contrasts_pivot = TE_contrasts %>%
                     #drop descriptor row
                     filter(L != "Pool1") %>%
                     #add treatment names to prepare for pivot
                     mutate(Treatment = rownames(.), .before = 1) %>%
                     pivot_longer(-Treatment,
                                  names_to = "contrast_name", values_to = "contrast_value") %>%
                     #concatenate contrast_name and treatment
                     mutate(contrast_name = str_c(contrast_name, "_", Treatment))
                     

contrast.matrix <- makeContrasts(contrasts = TE_contrasts_pivot$contrast_value, 
                      levels = design)

#rename matrix according to naming scheme above
contrast.matrix = contrast.matrix %>%
                  as.data.frame() %>% #convert to df for now for manipulation 
                                      #(use as.data.frame to preserve names)
                  rename_all(~plyr::mapvalues(., 
                                              from = TE_contrasts_pivot$contrast_value,
                                              to = TE_contrasts_pivot$contrast_name)) %>%
                  as.matrix()

fit.cont <- contrasts.fit(fit, contrast.matrix)

e.fit <- eBayes(fit.cont)

#first bind together all LFC lists for all genes from ea. contrast
log2FC.all = as.data.frame(map_dfc(1:length(colnames(e.fit$contrasts)),
                             function(coef) topTable(e.fit, coef=coef, number=Inf, adjust = "BH",
                                                     sort.by='none')$logFC)) %>%
      data.frame()

adj_p = as.data.frame(map_dfc(1:length(colnames(e.fit$contrasts)),
                                    function(coef) topTable(e.fit, coef=coef, adjust = "BH",
                                                            number=Inf,
                                                            sort.by='none')$adj.P.Val))

rownames(log2FC.all) = rownames(adj_p) = make.unique(rownames(dge_norm_fltd), sep = ".")
colnames(log2FC.all) = colnames(adj_p) = colnames(e.fit$contrasts)

#turn log2FC.all.binary into matrix of -1,0,1
FDR_thresh = 0.01
log2FC.all.binary = adj_p %>%
              mutate_all(~ifelse(. < FDR_thresh, 1, 0))

#filter LFC matrix accordingly
log2FC.all.binary[(log2FC.all < 0) == T] = -1 * log2FC.all.binary[(log2FC.all < 0) == T]
names = rownames(log2FC.all.binary)
log2FC.all.binary = log2FC.all.binary %>%
                    data.frame()

#only use adj p vals from TE contrasts
DE_log2FC.all = log2FC.all[rowSums(log2FC.all.binary %>% select(contains("TE")) != 0) > 0,] %>%
                     data.frame()
rownames(DE_log2FC.all) = rownames(log2FC.all)[rowSums(log2FC.all.binary %>% 
                                                         select(contains("TE")) != 0) > 0]

#filter DE matrix based on LFC thresh
LFC_thresh = 1
DE_LFC_thresh = DE_log2FC.all %>%
                     filter(if_any(contains("TE"), ~ abs(.) > LFC_thresh))

#write csvs of DEGs and all log2FC
write.csv(as.matrix(DE_LFC_thresh), path(output_dir, "TE_analysis/FDR0.01/FDR0.01_LFC1/20231107_filtered_FDR0.01_LFC1_TE-calcs_TE-DEGs-ONLY-lfcTable.csv"))

#save binary matrix
write.csv(as.matrix(log2FC.all.binary), path(output_dir, "TE_analysis/FDR0.01/FDR0.01_LFC1/20231108_filtered_FDR0.01_LFC1_TE-calcs_TE-binary-sig-matrix.csv"))
```

Second Approach: follows more closely approach found here: https://github.com/barnalab/regeneration/blob/main/10302022code.R
- Sum relevent pools (Light + medium or Medium + heavy) in raw data
- Normalize sums (cpm)
- Then compute contrasts
```{r}

#First add LM, MH and sudo-free cols to raw data 

```

Make pretty heatmaps!
```{r}
#set heatmap params----
#set seed for kmeans
set.seed(123)

#set fast clustering
ht_opt$fast_hclust = TRUE
ht_opt$message = FALSE

#prep data and make heatmaps----
#filter percent of transcript data by our 2883 sig. TE genes
#and remake heatmap
#4 genes were not DE but were TE
#"Trpm7"      "Spata6"     "Rpl36a.ps3" "Gm8355"  
#for the time being ill just remove since i dont think we care

#do one timepoint at at time for easier visualization
H_per = H[rownames(H) %in% rownames(DE_LFC_thresh), !(grepl("6h", colnames(H)))]
H_per_scale = H_per %>%
              t() %>%
              scale() %>%
              t()
topAnnot_per = top_annot %>%
               filter(Treatment != "LSG6h") %>%
               mutate(Pool = gsub("Pool1", "Light",
                      gsub("Pool2", "Medium",
                      gsub("Pool3", "Heavy", Pool))),
                      Pool = factor(Pool, 
                                    levels = c("Light", "Medium", "Heavy"))) %>%
               rename("Polysome Fraction" = Pool)

p_per <-Heatmap(H_per_scale,
        column_title = "% of Total mRNA in Pool (cpm)",
        column_title_gp = gpar(fontsize=12),
        name = "Avg_pool_cpm / Avg_input_cpm \n(row normalized)",
        show_row_names = F,
        show_column_names=F,
        #cluster_rows =T,
        cluster_columns = F,
        top_annotation = HeatmapAnnotation(df = topAnnot_per,
                         col = list(Treatment = c(Control = "grey",
                                                   LSG2h = "#F89F7C",
                                                  LSG6h = "#F68064"),
                                   `Polysome Fraction` = c("Light" = "#68C3A5", 
                                                          "Medium" ="#539C84",
                                                           "Heavy" ="#3E7563")),
                          border = T,
                          show_legend = F,
                          annotation_label = c("",""),
                          annotation_name_gp = gpar(fontsize = 8)),
        column_split = topAnnot_per$Treatment,
        column_gap = unit(0, "mm"),
        border = T,
        row_km = 20,
        row_title_gp = gpar(fontsize=7, angle = 0),
        row_gap = unit(0, "mm"),
        #border_gp = gpar(col = "black", lwd = 2), 
        col = colorRamp2(c(min(H_per_scale),-1, 0, max(H_per_scale)), 
                         c("white", "white","#B9CEE6","#8C9FCA"))
        )

#"#B9CEE6","#8C9FCA"
#remake LFC heatmap
#use LFC data we have now from TE analysis
H_LFC = DE_LFC_thresh %>%
        select(starts_with("L"), starts_with("M"), starts_with("H")) %>%
        select(contains("2h")) %>%
        as.matrix()
#reorder data to match
H_LFC = H_LFC[rownames(H_per),]
H_LFC_scale = H_LFC %>%
              t() %>%
              scale() %>%
              t()

topAnnot_LFC = data.frame(Name = colnames(H_LFC)) %>%
             separate(Name, into = c("Polysome Fraction", "Treatment"), sep = "_") %>%
             mutate(`Polysome Fraction` = gsub("L", "Light",
                                               gsub("M", "Medium", 
                                                    gsub("H", "Heavy", `Polysome Fraction`))),
                    `Polysome Fraction` = factor(`Polysome Fraction`, 
                                                 levels = c("Light", "Medium", "Heavy"))) %>%
             select(2,1) #reorder to match first heatmap
  

p_LFC <-Heatmap(H_LFC,
        column_title = "LFC",
        column_title_gp = gpar(fontsize=12),
        name = "log2(treated/control)",
        show_row_names = F,
        show_column_names=F,
        #cluster_rows =T,
        cluster_columns = F,
        top_annotation = HeatmapAnnotation(df = topAnnot_LFC,
                         col = list(Treatment = c(Control = "grey",
                                                   LSG2h = "#F89F7C",
                                                  LSG6h = "#F68064"),
                                   `Polysome Fraction` = c("Light" = "#68C3A5", 
                                                          "Medium" ="#539C84",
                                                          "Heavy" ="#3E7563")),
                          border = T,
                          show_legend = F,
                          annotation_label = c("",""),
                          annotation_name_gp = gpar(fontsize = 8)),
        column_split = topAnnot_LFC$Treatment,
        column_gap = unit(0, "mm"),
        row_km = 20,
        row_gap = unit(0, "mm"),
        border = T,
        #border_gp = gpar(col = "black", lwd = 2), 
        col = colorRamp2(c(-2.25,-0.3, 0, 0.3, 2.25), c("blue","white","white", "white", "red"))
        )
intermed_p <- p_per + p_LFC
p_LFC
intermed_p

#save intermediate heatmap
do_plot("TE_analysis/FDR0.01/FDR0.01_LFC1/complexHeatmap/percent-transcript_PLUS_LFC_2883genes_2h.pdf",
        8,6,intermed_p)

#make TE heatmap
H_TE = DE_LFC_thresh %>%
        select(contains("mRNA"), contains("TE")) %>%
        as.matrix()
#reorder data to match
H_TE = H_TE[rownames(H_per), !grepl("6h", colnames(H_TE))]

topAnnot_TE = data.frame(Name = colnames(H_TE)) %>%
              separate(Name, into = c("Metric", "Treatment"), sep = "_") %>%
              mutate(Metric = gsub("delta.mRNA", "\U0394mRNA",
                                   gsub("TE", "\U0394TE",
                                        Metric)),
                     Metric = factor(Metric, levels = c("\U0394mRNA", "\U0394TE"))) %>%
              select(2,1) # reoder so color blocks appear correctly

p_TE <- Heatmap(H_TE,
        column_title = "TE",
        column_title_gp = gpar(fontsize=12),
        name = "log2(((heavy)(medium))/light)",
        show_row_names = F,
        show_column_names=F,
        #cluster_rows =T,
        cluster_columns = F,
        top_annotation = HeatmapAnnotation(df = topAnnot_TE,
                         col = list(Treatment = c(Control = "grey",
                                                   LSG2h = "#F89F7C",
                                                  LSG6h = "#F68064"),
                                   Metric = c("\U0394mRNA" ="#D278AE",
                                              "\U0394TE" = "#AED29E")),
                          border = T,
                          show_legend = F,
                          annotation_label = c("",""),
                          annotation_name_gp = gpar(fontsize = 8)),
        column_split = topAnnot_TE$Treatment,
        column_gap = unit(0, "mm"),
        row_km = 20,
        row_gap = unit(0, "mm"),
        border = T,
        col = colorRamp2(c(-3,-1.5,-0.5 ,0,0.5, 1.5,3), c("darkblue", "#8C9FCA",
                                                            "white", "white", "white",
                                                           "orangered","firebrick")))

#D278AE
#AED29E
#put it all together
all3 <-p_per + p_LFC +p_TE
#all3

#first iteration plot save
do_plot("TE_analysis/FDR0.01/FDR0.01_LFC1/compiled_heatmap_2h_unordered.pdf", 8,6, all3)

#get clusters----
p_drawn <- draw(all3) #draw to lock in row clustering
p_drawn #run to show on screen and visually choose clusters

#use LFC to determine order instead of transcript distribution
p_LFC_drawn <-draw(p_LFC+p_TE)
p_LFC_drawn
rd <- row_order(p_LFC_drawn) #get row order & save

#get cluster gene lists
cluster_list = list() #initialize empty list
for(i in 1:20){
  cluster_list[[i]] = data.frame(Cluster = i, 
                                 gene = c(rownames(H_LFC)[ rd[[i]]]))
}
cluster_df = do.call(bind_rows, cluster_list)

#edit/condense clusters by eye
cluster_df_edit = cluster_df %>%
                  mutate(old_cluster = Cluster, 
                         Cluster = ifelse(old_cluster %in% c(16:20), 1,
                                    ifelse(old_cluster %in% c(11:13), 2,
                                    ifelse(old_cluster %in% c(10, 14:15), 3,
                                    ifelse(old_cluster %in% c(6,8:9), 4,
                                    ifelse(old_cluster %in% c(4,5,7), 5,
                                           6)))))) %>%
                  arrange(Cluster)
rownames(cluster_df_edit) = cluster_df_edit$gene

write.csv(as.matrix(cluster_df_edit), path(output_dir, "TE_analysis/FDR0.01/FDR0.01_LFC1/20231108_2h_compiledHeatmap_clusters.csv"))

#re-make heatmap with clusters----
#re-order all input matrices according to cluster order then re-draw
#also add left annotation with cluster info for splitting
gene_order = cluster_df_edit$gene
H_per_scale_reorder = H_per_scale[gene_order,]
H_LFC_reorder = H_LFC[gene_order,]
H_TE_reorder = H_TE[gene_order,]
leftAnnot = cluster_df_edit


#remake heatmaps
#clean up redudancies in code later and generalize!
p_per <-Heatmap(H_per_scale_reorder,
        column_title = "% of Total mRNA in Pool (cpm)",
        column_title_gp = gpar(fontsize=12),
        name = "Avg_pool_cpm / Avg_input_cpm \n(row normalized)",
        show_row_names = F,
        show_column_names=F,
        cluster_row_slices = T,
        #row_order = gene_order,
        cluster_rows =F,
        cluster_columns = F,
        top_annotation = HeatmapAnnotation(df = topAnnot_per,
                         col = list(Treatment = c(Control = "grey",
                                                   LSG2h = "#F89F7C",
                                                  LSG6h = "#F68064"),
                                   `Polysome Fraction` = c("Light" = "#68C3A5", 
                                                          "Medium" ="#539C84",
                                                           "Heavy" ="#3E7563")),
                          border = T,
                          show_legend = F,
                          annotation_label = c("",""),
                          annotation_name_gp = gpar(fontsize = 8)),
        column_split = topAnnot_per$Treatment,
        column_gap = unit(0, "mm"),
        border = T,
        #row_km = 20,
        row_split = leftAnnot$Cluster,
        row_title_gp = gpar(fontsize=7),
        row_gap = unit(0, "mm"),
        #border_gp = gpar(col = "black", lwd = 2), 
        col = colorRamp2(c(min(H_per_scale),-1, 0, max(H_per_scale)), 
                         c("white", "white","#B9CEE6","#707FA1"))
        )
p_per

p_LFC <-Heatmap(H_LFC_reorder,
        column_title = "LFC",
        column_title_gp = gpar(fontsize=12),
        name = "log2(treated/control)",
        show_row_names = F,
        show_column_names=F,
        cluster_rows =F,
        cluster_row_slices = T,
        #row_order = gene_order,
        cluster_columns = F,
        top_annotation = HeatmapAnnotation(df = topAnnot_LFC,
                         col = list(Treatment = c(Control = "grey",
                                                   LSG2h = "#F89F7C",
                                                  LSG6h = "#F68064"),
                                   `Polysome Fraction` = c("Light" = "#68C3A5", 
                                                          "Medium" ="#539C84",
                                                          "Heavy" ="#3E7563")),
                          border = T,
                          show_legend = F,
                          annotation_label = c("",""),
                          annotation_name_gp = gpar(fontsize = 8)),
        column_split = topAnnot_LFC$Treatment,
        column_gap = unit(0, "mm"),
        #row_km = 20,
        row_split = leftAnnot$Cluster,
        row_gap = unit(0, "mm"),
        border = T,
        #border_gp = gpar(col = "black", lwd = 2), 
        col = colorRamp2(c(-2.25,-0.3, 0, 0.3, 2.25), c("blue","white","white", "white", "red"))
        )
p_per + p_LFC

p_TE <- Heatmap(H_TE_reorder,
        column_title = "TE",
        column_title_gp = gpar(fontsize=12),
        name = "log2(((heavy)(medium))/light)",
        show_row_names = F,
        show_column_names=F,
        cluster_rows =F,
        cluster_row_slices = T,
        #row_order = gene_order,
        cluster_columns = F,
        top_annotation = HeatmapAnnotation(df = topAnnot_TE,
                         col = list(Treatment = c(Control = "grey",
                                                   LSG2h = "#F89F7C",
                                                  LSG6h = "#F68064"),
                                   Metric = c("\U0394mRNA" ="#D278AE",
                                              "\U0394TE" = "#AED29E")),
                          border = T,
                          show_legend = F,
                          annotation_label = c("",""),
                          annotation_name_gp = gpar(fontsize = 8)),
        column_split = topAnnot_TE$Treatment,
        column_gap = unit(0, "mm"),
        #row_km = 20,
        row_split = leftAnnot$Cluster,
        row_gap = unit(0, "mm"),
        border = T,
        col = colorRamp2(c(-3,-1.5,-0.5 ,0,0.5, 1.5,3), c("darkblue", "#8C9FCA",
                                                            "white", "white", "white",
                                                           "orangered","firebrick")))


all3_ordered = p_per + p_LFC + p_TE

#save plot
#she's beautiful i love her lets save it!
do_plot("TE_analysis/FDR0.01/FDR0.01_LFC1/20231108_2h_compiledHeatmap.pdf",8,6, all3_ordered)
```
Repeat but for 6h timepoint
```{r}

#prep data and make heatmaps----
#filter percent of transcript data by our 2883 sig. TE genes
#and remake heatmap
#4 genes were not DE but were TE
#"Trpm7"      "Spata6"     "Rpl36a.ps3" "Gm8355"  
#for the time being ill just remove since i dont think we care

#do one timepoint at at time for easier visualization
#2542 genes that are unique to 6h timepoint
H_per6 = H[rownames(H) %in% rownames(DE_LFC_thresh[rownames(DE_LFC_thresh) %in% rownames(log2FC.all.binary %>% 
                                                     data.frame() %>%
                                                     filter(TE_LSG6h != 0)),]), !(grepl("2h", colnames(H)))]
H_per_scale6 = H_per6 %>%
              t() %>%
              scale() %>%
              t()
topAnnot_per6 = top_annot %>%
               filter(Treatment != "LSG2h") %>%
               mutate(Pool = gsub("Pool1", "Light",
                      gsub("Pool2", "Medium",
                      gsub("Pool3", "Heavy", Pool))),
                      Pool = factor(Pool, 
                                    levels = c("Light", "Medium", "Heavy"))) %>%
               rename("Polysome Fraction" = Pool)

p_per6 <-Heatmap(H_per_scale6,
        column_title = "% of Total mRNA in Pool (cpm)",
        column_title_gp = gpar(fontsize=12),
        name = "Avg_pool_cpm / Avg_input_cpm \n(row normalized)",
        show_row_names = F,
        show_column_names=F,
        #cluster_rows =T,
        cluster_columns = F,
        top_annotation = HeatmapAnnotation(df = topAnnot_per,
                         col = list(Treatment = c(Control = "grey",
                                                   LSG2h = "#F89F7C",
                                                  LSG6h = "#F68064"),
                                   `Polysome Fraction` = c("Light" = "#68C3A5", 
                                                          "Medium" ="#539C84",
                                                           "Heavy" ="#3E7563")),
                          border = T,
                          show_legend = F,
                          annotation_label = c("",""),
                          annotation_name_gp = gpar(fontsize = 8)),
        column_split = topAnnot_per6$Treatment,
        column_gap = unit(0, "mm"),
        border = T,
        row_km = 20,
        row_title_gp = gpar(fontsize=7, angle = 0),
        row_gap = unit(0, "mm"),
        #border_gp = gpar(col = "black", lwd = 2), 
        col = colorRamp2(c(min(H_per_scale),-1, 0, max(H_per_scale)), 
                         c("white", "white","#B9CEE6","#8C9FCA"))
        )

#"#B9CEE6","#8C9FCA"
#remake LFC heatmap
#use LFC data we have now from TE analysis
H_LFC6 = DE_LFC_thresh %>%
        select(starts_with("L"), starts_with("M"), starts_with("H")) %>%
        select(contains("6h")) %>%
        as.matrix()

#reorder data to match
H_LFC6 = H_LFC6[rownames(H_per6),]
H_LFC_scale = H_LFC %>%
              t() %>%
              scale() %>%
              t()

topAnnot_LFC6 = data.frame(Name = colnames(H_LFC6)) %>%
             separate(Name, into = c("Polysome Fraction", "Treatment"), sep = "_") %>%
             mutate(`Polysome Fraction` = gsub("L", "Light",
                                               gsub("M", "Medium", 
                                                    gsub("H", "Heavy", `Polysome Fraction`))),
                    `Polysome Fraction` = factor(`Polysome Fraction`, 
                                                 levels = c("Light", "Medium", "Heavy"))) %>%
             select(2,1) #reorder to match first heatmap
  

p_LFC6 <-Heatmap(H_LFC6,
        column_title = "LFC",
        column_title_gp = gpar(fontsize=12),
        name = "log2(treated/control)",
        show_row_names = F,
        show_column_names=F,
        #cluster_rows =T,
        cluster_columns = F,
        top_annotation = HeatmapAnnotation(df = topAnnot_LFC,
                         col = list(Treatment = c(Control = "grey",
                                                   LSG2h = "#F89F7C",
                                                  LSG6h = "#F68064"),
                                   `Polysome Fraction` = c("Light" = "#68C3A5", 
                                                          "Medium" ="#539C84",
                                                          "Heavy" ="#3E7563")),
                          border = T,
                          show_legend = F,
                          annotation_label = c("",""),
                          annotation_name_gp = gpar(fontsize = 8)),
        column_split = topAnnot_LFC6$Treatment,
        column_gap = unit(0, "mm"),
        row_km = 20,
        row_gap = unit(0, "mm"),
        border = T,
        #border_gp = gpar(col = "black", lwd = 2), 
        col = colorRamp2(c(-2.25,-0.3, 0, 0.3, 2.25), c("blue","white","white", "white", "red"))
        )
intermed_p6 <- p_per6 + p_LFC6
#p_LFC
intermed_p6

#save intermediate heatmap
do_plot("TE_analysis/FDR0.01/FDR0.01_LFC1/complexHeatmap/percent-transcript_PLUS_LFC_2542genes_6h.pdf",
        8,6,intermed_p6)

#make TE heatmap
H_TE6 = DE_LFC_thresh %>%
        select(contains("mRNA"), contains("TE")) %>%
        as.matrix()
#reorder data to match
H_TE6 = H_TE6[rownames(H_per6), !grepl("2h", colnames(H_TE6))]

topAnnot_TE6 = data.frame(Name = colnames(H_TE6)) %>%
              separate(Name, into = c("Metric", "Treatment"), sep = "_") %>%
              mutate(Metric = gsub("delta.mRNA", "\U0394mRNA",
                                   gsub("TE", "\U0394TE",
                                        Metric)),
                     Metric = factor(Metric, levels = c("\U0394mRNA", "\U0394TE"))) %>%
              select(2,1) # reoder so color blocks appear correctly

p_TE6 <- Heatmap(H_TE6,
        column_title = "TE",
        column_title_gp = gpar(fontsize=12),
        name = "log2(((heavy)(medium))/light)",
        show_row_names = F,
        show_column_names=F,
        #cluster_rows =T,
        row_title_gp = gpar(fontsize = 7),
        cluster_columns = F,
        top_annotation = HeatmapAnnotation(df = topAnnot_TE,
                         col = list(Treatment = c(Control = "grey",
                                                   LSG2h = "#F89F7C",
                                                  LSG6h = "#F68064"),
                                   Metric = c("\U0394mRNA" ="#D278AE",
                                              "\U0394TE" = "#AED29E")),
                          border = T,
                          show_legend = F,
                          annotation_label = c("",""),
                          annotation_name_gp = gpar(fontsize = 8)),
        column_split = topAnnot_TE6$Treatment,
        column_gap = unit(0, "mm"),
        row_km = 20,
        row_gap = unit(0, "mm"),
        border = T,
        col = colorRamp2(c(-3,-1.5,-0.5 ,0,0.5, 1.5,3), c("darkblue", "#8C9FCA",
                                                            "white", "white", "white",
                                                           "orangered","firebrick")))

#D278AE
#AED29E
#put it all together
all3_6h <-p_per6 + p_LFC6 +p_TE6
#all3

#first iteration plot save
do_plot("TE_analysis/FDR0.01/FDR0.01_LFC1/compiled_heatmap_2h_unordered.pdf", 8,6, all3)

#get clusters----
p_drawn6 <- draw(all3_6h) #draw to lock in row clustering
p_drawn6 #run to show on screen and visually choose clusters

#use LFC to determine order instead of transcript distribution
p_LFC_drawn6 <-draw(p_TE6+p_LFC6)
p_LFC_drawn6
rd <- row_order(p_LFC_drawn6) #get row order & save

#get cluster gene lists
cluster_list6 = list() #initialize empty list
for(i in 1:20){
  cluster_list6[[i]] = data.frame(Cluster = i, 
                                 gene = c(rownames(H_LFC6)[ rd[[i]]]))
}
cluster_df_6h = do.call(bind_rows, cluster_list6)

#edit/condense clusters by eye
cluster_df_edit_6h = cluster_df_6h %>%
                  mutate(old_cluster = Cluster, 
                         Cluster = ifelse(old_cluster %in% c(17:20), 1,
                                    ifelse(old_cluster %in% c(15:16), 2,
                                    ifelse(old_cluster %in% c(14), 3,
                                    ifelse(old_cluster %in% c(13), 4,
                                    ifelse(old_cluster %in% c(12), 6,
                                    ifelse(old_cluster %in% c(10:11), 7,
                                    ifelse(old_cluster %in% c(8:9), 9,
                                    ifelse(old_cluster %in% c(6,7), 10, 
                                    ifelse(old_cluster %in% c(5,2), 11, 
                                           12)))))))))) %>%
                  arrange(Cluster)
rownames(cluster_df_edit_6h) = cluster_df_edit_6h$gene

write.csv(as.matrix(cluster_df_edit_6h), path(output_dir, "TE_analysis/FDR0.01/FDR0.01_LFC1/20231109_6h_compiledHeatmap_clusters_12Clusters.csv"))

#re-make heatmap with clusters----
#re-order all input matrices according to cluster order then re-draw
#also add left annotation with cluster info for splitting
gene_order6 = cluster_df_edit_6h$gene
H_per_scale_reorder6 = H_per_scale6[gene_order6,]
H_LFC_reorder6 = H_LFC6[gene_order6,]
H_TE_reorder6 = H_TE6[gene_order6,]
leftAnnot6 = cluster_df_edit_6h


#remake heatmaps
#clean up redudancies in code later and generalize!
p_per6 <-Heatmap(H_per_scale_reorder6,
        column_title = "% of Total mRNA in Pool (cpm)",
        column_title_gp = gpar(fontsize=12),
        name = "Avg_pool_cpm / Avg_input_cpm \n(row normalized)",
        show_row_names = F,
        show_column_names=F,
        cluster_row_slices = T,
        #row_order = gene_order,
        cluster_rows =F,
        cluster_columns = F,
        top_annotation = HeatmapAnnotation(df = topAnnot_per6,
                         col = list(Treatment = c(Control = "grey",
                                                   LSG2h = "#F89F7C",
                                                  LSG6h = "#F68064"),
                                   `Polysome Fraction` = c("Light" = "#68C3A5", 
                                                          "Medium" ="#539C84",
                                                           "Heavy" ="#3E7563")),
                          border = T,
                          show_legend = F,
                          annotation_label = c("",""),
                          annotation_name_gp = gpar(fontsize = 8)),
        column_split = topAnnot_per6$Treatment,
        column_gap = unit(0, "mm"),
        border = T,
        #row_km = 20,
        row_split = leftAnnot6$Cluster,
        row_title_gp = gpar(fontsize=7),
        row_gap = unit(0, "mm"),
        #border_gp = gpar(col = "black", lwd = 2), 
        col = colorRamp2(c(min(H_per_scale),-1, 0, max(H_per_scale)), 
                         c("white", "white","#B9CEE6","#707FA1"))
        )
p_per6

p_LFC6 <-Heatmap(H_LFC_reorder6,
        column_title = "LFC",
        column_title_gp = gpar(fontsize=12),
        name = "log2(treated/control)",
        show_row_names = F,
        show_column_names=F,
        cluster_rows =F,
        cluster_row_slices = T,
        #row_order = gene_order,
        cluster_columns = F,
        top_annotation = HeatmapAnnotation(df = topAnnot_LFC6,
                         col = list(Treatment = c(Control = "grey",
                                                   LSG2h = "#F89F7C",
                                                  LSG6h = "#F68064"),
                                   `Polysome Fraction` = c("Light" = "#68C3A5", 
                                                          "Medium" ="#539C84",
                                                          "Heavy" ="#3E7563")),
                          border = T,
                          show_legend = F,
                          annotation_label = c("",""),
                          annotation_name_gp = gpar(fontsize = 8)),
        column_split = topAnnot_LFC6$Treatment,
        column_gap = unit(0, "mm"),
        #row_km = 20,
        row_split = leftAnnot6$Cluster,
        row_gap = unit(0, "mm"),
        border = T,
        #border_gp = gpar(col = "black", lwd = 2), 
        col = colorRamp2(c(-2.25,-0.3, 0, 0.3, 2.25), c("blue","white","white", "white", "red"))
        )
p_per6 + p_LFC6

p_TE6 <- Heatmap(H_TE_reorder6,
        column_title = "TE",
        column_title_gp = gpar(fontsize=12),
        name = "log2(((heavy)(medium))/light)",
        show_row_names = F,
        show_column_names=F,
        cluster_rows =F,
        cluster_row_slices = T,
        #row_order = gene_order,
        cluster_columns = F,
        top_annotation = HeatmapAnnotation(df = topAnnot_TE6,
                         col = list(Treatment = c(Control = "grey",
                                                   LSG2h = "#F89F7C",
                                                  LSG6h = "#F68064"),
                                   Metric = c("\U0394mRNA" ="#D278AE",
                                              "\U0394TE" = "#AED29E")),
                          border = T,
                          show_legend = F,
                          annotation_label = c("",""),
                          annotation_name_gp = gpar(fontsize = 8)),
        column_split = topAnnot_TE6$Treatment,
        column_gap = unit(0, "mm"),
        #row_km = 20,
        row_split = leftAnnot6$Cluster,
        row_gap = unit(0, "mm"),
        border = T,
        col = colorRamp2(c(-3,-1.5,-0.5 ,0,0.5, 1.5,3), c("darkblue", "#8C9FCA",
                                                            "white", "white", "white",
                                                           "orangered","firebrick")))


all3_ordered_6h = p_per6 + p_LFC6 + p_TE6

p_LFC + p_TE + p_LFC6 + p_TE6
#save plot
#she's beautiful i love her lets save it!
do_plot("TE_analysis/FDR0.01/FDR0.01_LFC1/ComplexHeatmap/6h/20231109_6h_compiledHeatmap_12Clusters.pdf",8,6, all3_ordered_6h)
```


#mRNA v TE correlation plot
```{r}
comparison_df = log2FC.all %>%
                select(contains("mRNA"), contains("TE")) %>%
                rownames_to_column("gene") %>%
                pivot_longer(-gene, names_to = "metric", values_to = "LFC") %>%
                separate(metric, into = c("metric", "Treatment"), sep = "_") %>%
                mutate(metric = gsub("delta.mRNA", "LFC mRNA",
                                     gsub("TE", "LFC TE", metric))) %>%
                pivot_wider(names_from = "metric", values_from = "LFC")

quick_plot <-ggplot(comparison_df, aes(x = `LFC mRNA`, y = `LFC TE`)) +
  geom_point() +
  geom_point(data = subset(comparison_df, abs(`LFC mRNA`) < 1 & abs(`LFC TE`) > 1), aes(x = `LFC mRNA`, y = `LFC TE`),
             color = "#AED29E")+
  facet_wrap(~Treatment, scales = "free")


save_plot(path(output_dir, "TE_analysis/FDR0.01/FDR0.01_LFC1/mRNA-v-TE-correlation_test_plot_all_genes.pdf"), quick_plot)

#get group of genes that are interesting for follow-up
int_genes <- subset(comparison_df, abs(`LFC mRNA`) < 1 & abs(`LFC TE`) > 1)
int_genes_up <- subset(comparison_df, abs(`LFC mRNA`) < 1 & `LFC TE` > 1)

mini_binary <- cbind(gene = rownames(log2FC.all.binary), log2FC.all.binary %>% select(contains("TE")))
int_genes_sig <- int_genes_up %>%
                 left_join(mini_binary, by = "gene") %>%
                 filter(!(TE_LSG2h == 0 & TE_LSG6h == 0)) %>%
                 rename_at(c(5:6), ~str_c(., "_sig")) %>%
                 mutate_at(c(5:6), ~as.logical(.)) %>%
                 arrange(gene)

write.csv(as.matrix(int_genes_sig), path(output_dir, "TE_analysis/FDR0.01/FDR0.01_LFC1/TE_up_mRNA_noChange_genes_LFC1_sigFDR0.01.csv"))
```

